src/Workspaces/Core/Portable/Shared/Extensions/ISolutionExtensions.cs

```diff
@@ -38,6 +39,20 @@ internal static partial class ISolutionExtensions
         public static TextDocumentKind? GetDocumentKind(this Solution solution, DocumentId documentId)
             => solution.GetTextDocument(documentId)?.Kind;
 
+        internal static TextDocument? GetTextDocumentForLocation(this Solution solution, Location location)
+        {
+            switch (location.Kind)
+            {
+                case LocationKind.SourceFile:
+                    return solution.GetDocument(location.SourceTree);
+                case LocationKind.ExternalFile:
+                    var documentId = solution.GetDocumentIdsWithFilePath(location.GetLineSpan().Path).FirstOrDefault();
```

Comment:
Why not `Single` here?

Reply:
There can be multiple documents (linked files) or no documents matching the file path. Analyzers can report a diagnostic with any non-existent file path.

Reply:
Out of scope for this pr: I find the idea of an analyzer reporting a diagnostic for a file path not in the compilation upsetting :)